### PR TITLE
feat: solr health check before search

### PR DESCRIPTION
**Pull request by yoyoyohamapi**
I found if our solr server is not available, users would not be able to visit any topic either. The problem was caused by the *Suggested Topics Widget* in [nodebb-widget-essentials](nodebb-widget-essentials) plugin and we activated it in topic template, this widget will trigger a search if our search engine enabled: https://github.com/NodeBB/NodeBB/blob/master/src/topics/suggested.js#L72.

So before we do a search, we may need check service health first. @julianlam @Dravere

### CODE REVIEW

File: library.js [hunk 1]
```diff
@@ -182,23 +182,31 @@ Solr.search = function (data, callback) {
...
-				return callback(err);
-			} else if (obj && obj.response && obj.response.docs.length > 0) {
-				const payload = obj.response.docs.map(function (result) {
-					return result[field];
-				}).filter(Boolean);
-
-				callback(null, payload);
-				cache.set(term, payload);
+				// Stop use search engine when connection failed
+				winston.error('[plugins/solr] Could not connect to Solr server' );
```
Comment (julianlam) [inline library.js]: Assuming Solr server being down is a transient issue, perhaps `winston.warn` would be better?

File: library.js [hunk 2]
```diff
@@ -182,23 +182,31 @@ Solr.search = function (data, callback) {
 
 		const query = Solr.client.createQuery().q(term).edismax().qf(fields).start(0).rows(500);
 
-		Solr.client.search(query, function (err, obj) {
+		Solr.ping(function(err) {
```
Comment (julianlam) [inline library.js]: Hey, can you maybe do `await Solr.ping()` instead? You'd have to promisify `Solr.ping`, or alternatively rewrite it as an asynchronous function :thinking:

File: library.js (hunk 1 above)
Comment (yoyoyohamapi) [reply library.js]: @julianlam :ok_hand:

File: library.js (hunk 2 above)
Comment (yoyoyohamapi) [reply library.js]: @yoyoyohamapi Solr ping is already an asynchronous function: https://github.com/julianlam/nodebb-plugin-solr/blob/bd68f72e00bcd19c36e7d5e22fffffcb8685f64d/library.js#L71

File: library.js [hunk 3]
```diff
@@ -185,7 +185,7 @@ Solr.search = function (data, callback) {
 		Solr.ping(function(err) {
 			if (err) {
 				// Stop use search engine when connection failed
-				winston.error('[plugins/solr] Could not connect to Solr server' );
+				winston.warn('[plugins/solr] Could not connect to Solr server' );
```
Comment (yoyoyohamapi) [inline library.js]: @julianlam replace `winston.error` with `winston.warn`

File: library.js (hunk 2 above)
Comment (yoyoyohamapi) [reply library.js]: @julianlam any idea of this?

Comment (julianlam) [reply library.js]: I mean you'd have to write it as an `async function`. Don't worry, you can leave it alone then.

**Comment by yoyoyohamapi** [comment]
@Dravere Could you update the npm pkg version for this feature? Thanks a lot

**Comment by julianlam** [comment]
v5.2.1

**Pull request merged**